File: examples/python/afl_odds.py
"""
PuntersEdge API - AFL Odds Example
Fetch AFL head-to-head odds and find the best available price per team.

Get a free API key (1,500 credits/month, no credit card):
    https://puntersedge.online/developers/getting-started
API docs: https://puntersedge.online/developers/sports-odds-api-australia

Cost: /v1/sports/{sport_key}/odds bills 1 credit per requested market, so the
single markets=h2h call below costs 1 credit.
"""

import json
import sys

import requests

API_KEY = "your_api_key_here"
BASE_URL = "https://api.puntersedge.online/v1"

headers = {"X-API-Key": API_KEY}


def get_afl_odds():
    """Fetch current AFL head-to-head odds.

    The sport is a PATH segment (/sports/afl/odds), and the query parameter is
    `markets` - plural, comma-separated. Other valid params: bookmakers,
    competition, include_unknown_competition, oddsFormat, maxAgeMinutes.

    Racing is NOT a sport_key. For horse, harness or greyhound racing use
    /v1/racing/next-to-go instead - see horse_racing_ntg.py.

    Returns a bare JSON array of events. There is no "data" or "events" wrapper.
    """
    resp = requests.get(
        f"{BASE_URL}/sports/afl/odds",
        params={"markets": "h2h"},  # oddsFormat defaults to decimal
        headers=headers,
        timeout=10,
    )
    if resp.status_code in (401, 402, 429):
        # Errors are RFC 7807 application/problem+json: type, title, status, detail.
        problem = resp.json()
        # `detail` is a string for most errors, an ARRAY of field errors on 422 and an
        # OBJECT on 429 — print it defensively rather than interpolating a raw dict.
        _d = problem.get('detail')
        if not isinstance(_d, str):
            _d = json.dumps(_d)
        sys.exit(f"{resp.status_code} {problem.get('title')}: {_d}")
    resp.raise_for_status()
    return resp.json()


def _tokens(name):
    """Lowercase word tokens of a team name, punctuation stripped."""
    cleaned = "".join(c if c.isalnum() else " " for c in name)
    return {t for t in cleaned.lower().split() if t}


def match_team(outcome_name, home, away):
    """Map one bookmaker's outcome name onto the event's home or away team.

    Bookmakers do not agree on team names and the API passes each book's own
    spelling through untouched: on a live AFL event one book priced
    "Gold Coast" while another priced "Gold Coast SUNS". Grouping on the raw
    name splits one team into two rows and understates its best price.

    Matches when one token set contains the other. Returns None when the name
    is ambiguous or unrecognised - reporting it unmatched beats assigning the
    price to the wrong side.
    """
    got = _tokens(outcome_name)
    hits = [
        team
        for team in (home, away)
        if team and got and (got <= _tokens(team) or _tokens(team) <= got)
    ]
    return hits[0] if len(hits) == 1 else None


def best_price(event):
    """Best available h2h price per team across every bookmaker on the event.

    Returns (matched, unmatched): both map a label to {"price", "bookmaker"}.
    """
    home, away = event.get("home_team"), event.get("away_team")
    matched, unmatched = {}, {}
    for bookmaker in event.get("bookmakers", []):
        bk_name = bookmaker["key"]
        for market in bookmaker.get("markets", []):
            if market["key"] != "h2h":
                continue
            for outcome in market.get("outcomes", []):
                price = outcome["price"]
                team = match_team(outcome["name"], home, away)
                target = matched if team else unmatched
                label = team or outcome["name"]
                if label not in target or price > target[label]["price"]:
                    target[label] = {"price": price, "bookmaker": bk_name}
    return matched, unmatched


def main():
    events = get_afl_odds()
    print(f"Found {len(events)} AFL events\n")

    for event in events:
        home = event.get("home_team") or "Home"
        away = event.get("away_team") or "Away"
        print(f"{home} vs {away}  |  {event['commence_time']}")

        # Coverage on AFL is thin: at most 5 bookmakers price an event and some
        # carry three, so "best price" here means best of a handful, not of
        # the whole Australian market.
        books = [b["key"] for b in event.get("bookmakers", [])]
        print(f"  {len(books)} bookmaker(s): {', '.join(books) or 'none'}")

        # `stale` is true when the OLDEST contributing book has not updated in
        # 30 minutes (twice the 15-minute sports poll); `stale_bookmakers`
        # names them so you can drop those prices and keep the rest.
        if event.get("stale"):
            stale_books = ", ".join(event.get("stale_bookmakers", [])) or "unknown"
            print(f"  STALE - no recent update from: {stale_books}")

        matched, unmatched = best_price(event)
        for team, info in matched.items():
            print(f"  {team:30s}  Best: ${info['price']:.2f}  ({info['bookmaker']})")
        for name, info in unmatched.items():
            print(f"  {name:30s}  ${info['price']:.2f}  ({info['bookmaker']}) [unmatched name]")
        print()

    if not events:
        print("No upcoming AFL events. The AFL season runs March to September;")
        print("try markets=h2h on nba or nfl out of season — nrl runs Mar-Oct, alongside AFL.")


if __name__ == "__main__":
    main()

File: examples/javascript/afl_odds.js
/**
 * PuntersEdge API — AFL Odds (browser / Node 18+ fetch)
 * Fetch AFL head-to-head odds and display the best price per team.
 *
 * Free tier: 1,500 credits per month, no credit card. Credits are not requests —
 * this endpoint bills 1 credit per requested market, so one h2h call costs 1.
 *
 * Get a free API key: https://puntersedge.online/developers/getting-started
 * API docs: https://puntersedge.online/developers/sports-odds-api-australia
 */

const API_KEY = "your_api_key_here";
const BASE_URL = "https://api.puntersedge.online/v1";

async function getAflOdds() {
  // The sport is a PATH segment, and the query parameter is `markets` (plural,
  // comma-separated: h2h,spreads,totals). Other valid parameters: bookmakers,
  // competition, include_unknown_competition, oddsFormat, maxAgeMinutes.
  // Racing is NOT a sport_key — it lives under /v1/racing/* (e.g. next-to-go).
  const url = new URL(`${BASE_URL}/sports/afl/odds`);
  url.searchParams.set("markets", "h2h");

  const res = await fetch(url.toString(), {
    headers: { "X-API-Key": API_KEY },
  });

  if (!res.ok) {
    // Errors are RFC 7807 problem+json: { type, title, status, detail }.
    // 401 = bad/missing key, 402 = monthly credits exhausted,
    // 429 = rate limited (honour the Retry-After header, in seconds).
    let note = "";
    try {
      const problem = await res.json();
      const detail = typeof problem.detail === "string" ? problem.detail : problem.title;
      if (detail) note = ` — ${detail}`;
    } catch (_) {
      /* body was not JSON */
    }
    throw new Error(`API error ${res.status}${note}`);
  }

  // This endpoint returns a BARE JSON array of events. There is no { data: ... }
  // wrapper, so do not reach for one.
  return res.json();
}

/**
 * Fold a bookmaker's outcome name onto the event's canonical home_team/away_team.
 * Books do not agree on team names: on a live AFL card TAB sends "Brisbane" and
 * "Gold Coast" where Sportsbet sends "Brisbane Lions" and "Gold Coast SUNS", so
 * keying on the raw name splits one team into two rows. Case-insensitive prefix
 * matching in either direction handles those. It does NOT handle abbreviations
 * that share no prefix — "GWS"/"GWS GIANTS" will not fold onto "Greater Western
 * Sydney", nor "Wst Bulldogs" onto "Western Bulldogs". Names that fail to match
 * are kept under their raw spelling rather than dropped; a production app wants
 * a real alias table.
 */
function canonicalTeam(event, name) {
  const raw = name.toLowerCase();
  for (const team of [event.home_team, event.away_team]) {
    if (!team) continue;
    const canon = team.toLowerCase();
    if (raw.startsWith(canon) || canon.startsWith(raw)) return team;
  }
  return name;
}

function bestPrices(event) {
  // Event shape: { id, sport_key, sport_title, competition, commence_time,
  // home_team, away_team, odds_format, bookmakers[], data_age_seconds,
  // freshest_age_seconds, stale, stale_bookmakers[], data_quality, ... }
  // Each bookmaker: { key, title, last_update, age_seconds, stale, quality,
  // markets[] } and each market: { key, quality, outcomes[{ name, price }] }.
  const best = {};
  for (const bk of event.bookmakers || []) {
    for (const market of bk.markets || []) {
      if (market.key !== "h2h") continue;
      for (const outcome of market.outcomes || []) {
        const team = canonicalTeam(event, outcome.name);
        const price = outcome.price;
        if (typeof price !== "number") continue;
        if (!best[team] || price > best[team].price) {
          best[team] = { price, bookmaker: bk.key };
        }
      }
    }
  }
  return best;
}

async function main() {
  const events = await getAflOdds();
  console.log(`Found ${events.length} AFL events\n`);

  for (const event of events) {
    // commence_time is UTC ISO 8601. Coverage on sports is thinner than on
    // racing: AFL h2h currently comes from about five books (betright,
    // palmerbet, pointsbetau, sportsbet, tab), spreads from three, and some
    // sports carry only one — so "best price" here means best of what is live,
    // not best of the whole market.
    const bookCount = (event.bookmakers || []).length;
    console.log(`${event.home_team} vs ${event.away_team}  |  ${event.commence_time}`);
    console.log(
      `  ${bookCount} bookmaker(s)` +
        // data_age_seconds is the age of the OLDEST contributing price, not the
        // newest; stale_bookmakers names the books that have stopped updating.
        (event.data_age_seconds != null ? `, oldest price ${event.data_age_seconds}s old` : "") +
        (event.stale ? "  [STALE]" : "") +
        (event.stale_bookmakers && event.stale_bookmakers.length
          ? `  stale: ${event.stale_bookmakers.join(", ")}`
          : "")
    );

    const prices = bestPrices(event);
    for (const [team, { price, bookmaker }] of Object.entries(prices)) {
      console.log(`  ${team.padEnd(30)} Best: $${price.toFixed(2)}  (${bookmaker})`);
    }
    console.log();
  }
}

main().catch((err) => console.error(err.message));
